### PR TITLE
[Merged by Bors] - refactor (NumberTheory/ModularForms, Analysis/Complex/UpperHalfPlane) Refactor and clean up coercions

### PR DESCRIPTION
**Pull request by CBirkbeck**
The code for SL(2, Z) and modular forms had lots of custom local notations, repeated in each file, for certain chains of coercions. We get rid of many of these, and harmonise the remaining ones using scoped (rather than file-local) notations. Separately, we get rid of a lot of annoying duplicate API lemmas by avoiding using subgroups as subtypes unless absolutely necessary.

Co-authored-by: David Loeffler <d.loeffler.01@cantab.net>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean [hunk 1]
```diff
@@ -29,10 +29,10 @@ we state lemmas in this file without spurious `coe_fn` terms. -/
 attribute [-instance] Matrix.SpecialLinearGroup.instCoeFun
 attribute [-instance] Matrix.GeneralLinearGroup.instCoeFun
 
-local notation "GL(" n ", " R ")" "⁺" => Matrix.GLPos (Fin n) R
-local notation:1024 "↑ₘ" A:1024 =>
+local notation:1024 "↑ₘₙ" A:1024 =>
```
Comment (loefflerd) [inline Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean:32]: I still find this very confusing. We seem to have _local_ notations `↑ₘₙ` and `↑ₘₙ[R]` defined here, and also _scoped_ (non-local) notations with the same symbols in `SlashActions.lean` but i can't quite tell if they're actually the same; and we now have `↑ₘ` (without `n`) as well, defined in yet another file.

Can you add docstrings to the ones which don't currently have them, and cross-reference between the docstrings to explain how they are different?

Comment (CBirkbeck) [reply Mathlib/Analysis/Complex/UpperHalfPlane/Basic.lean:32]: Ah yes I should add the doc string here also. It seems we had `↑ₘ` meaning at least 2 different things, one was coercing through SL to matrices and the other was via GL_pos to GL, to matrices. I left the former as ↑ₘ and the latter is now ↑ₘₙ (the mn for many coercions I guess as it wouldnt give me a small g). I made them scoped in the ModularForms and ModularGroup namespaces.

**Comment by loefflerd** [comment]
I think we're moving in the right direction here. It's still annoying that we need these up-arrow notations at all, but at least this imposes a bit more order upon them. 

I have two further refactoring suggestions:

- migrate the last 150 lines of `LinearAlgebra/Matrix/SpecialLinearGroup.lean` (the part which is specific to SL2Z) into `Modular.lean`, which avoids having two different definitions of the same notation in the same file
- rename `Modular.lean` to `ModularGroup.lean`
- for each scoped notation, say in the docstring for it what namespace it is scoped in (since the mouseover text in VSCode doesn't provide this information automatically)

**Comment by loefflerd** [comment]
!bench

**Comment by loefflerd** [comment]
!bench

**Review by jcommelin — COMMENTED** [review]
LGTM

bors d=loefflerd

**Pull request closed without merging**